### DISCUSSION TITLE
Should someone write an article on medium?

### DISCUSSION THREAD

**Discussion by unicomp21** [esfx/esfx#66] [Ideas]
We have the most amazing typescript lib on the planet, and nobody knows it exists?

**Comment by hinell** [comment]
Moreover, almost nobody knows why it exists and even how to use it if there is chance of ever-existing use-case for it.

Would be best to provide examples on how to use it. Some examples are puzzling.

**Reply by rbuckton** [reply]
I can always update the docs with better examples if there are some areas you could suggest need improvement.

**Reply by hinell** [reply]
Would be nice to have a general overview of the library and its purpose. I came across it and still have no idea what these collections and rest things are for. Except probably HashMap etc. 

Thanks!

**Reply by rbuckton** [reply]
The entire `esfx` project is based on all of the things I'd like to see ECMAScript's standard library/built-ins support in the future, in some form or another. Most packages are either based on existing proposal, forthcoming proposals, or are experiments intended to prove out certain behaviors that I may pursue as a future proposal.

Many of the packages are geared towards defining low-level "protocols" (roughly analogous to what is suggested by https://github.com/tc39/proposal-first-class-protocols) for common capabilities, such as:

- `@esfx/equatable` - A symbol-based protocol for defining equality and comparability
- `@esfx/collection-core` - common collection protocols
- `@esfx/disposable` - common resource management protocols, per the [Explicit Resource Management Proposal](https://github.com/tc39/proposal-explicit-resource-management) (Stage 3)
- `@esfx/cancelable` - a protocol for cancellation, based on the [Cancellation Proposal](https://github.com/tc39/proposal-cancellation)
- `@esfx/async-lockable` - a protocol for locking for async coordination primitives (possible future proposal)
- `@esfx/threading-lockable` - a protocol for thread coordination primitives (related to the [Structs and Shared Structs Proposal](https://github.com/tc39/proposal-structs)).

The `HashMap` and `HashSet` collection classes are examples of how user-defined equality via the `equatable` package can provide flexible collections that don't rely only on reference-equality, such as creating a `HashMap` with `URI` keys, rather than strings.

The `SortedMap` and `SortedSet` collection classes are examples of how user-defined relational comparisons via the `equatable` package can provide collections whose contents are pre-sorted on insert.

Awhile back I also added my old `iterable-query` package to the `@esfx` namespace to investigate a counter-point to what is now the [Iterator Helpers Proposal](https://github.com/tc39/proposal-iterator-helpers), though it has more in common with Lodash/Underscore/Ramda:
- `@esfx/fn` - A number of utility functions to benefit developers using a functional programming (FP) style.
- `@esfx/iter-fn` - iteration functions a la Lodash/Underscore/Ramda, geared towards FP-style
- `@esfx/iter-query` - a method-chaining based query/iteration API that turns the FP-style functions in `iter-fn` into chainable methods.
- `@esfx/async-iter-fn` - async iteration functions a la Lodash/Underscore/Ramda, geared towards FP-style
- `@esfx/async-iter-query` - a method-chaining based async query/iteration API that turns the FP-style functions in `async-iter-fn` into chainable methods.
  
The various `@esfx/async-*` packages provide async coordination primitives that can be very useful when writing concurrent JavaScript that relies on `Promise` or `async`/`await`. When users write async JavaScript, there is the potential for data races and deadlocks even in a single-user application. Classes like `AsyncSemaphore` limit the number of concurrent operations that can access a shared resource. 

`@esfx/async-deferred` is an API that is analogous to the [`Promise.withResolvers()` Proposal](https://github.com/tc39/proposal-promise-with-resolvers) (now Stage 3 as of last week's TC39 plenary), and `@esfx/async-canceltoken` is an implementation of the `@esfx/cancelable` protocol and implements the API originally proposed in the [Cancellation Proposal](https://github.com/tc39/proposal-cancellation).

The thread synchronization primitives in `@esfx/threading-*` are multi-threading versions of the `@esfx/async-*` primitives that are designed to be used with Web Workers and `SharedArrayBuffer`, and a variation of `@esfx/threading-mutex` and `@esfx/threading-conditionvariable` is currently under discussion as part of the [Structs and Shared Structs Proposal](https://github.com/tc39/proposal-structs).

Speaking of which, the `@esfx/struct-type` package was inspired by an early version of an old typed objects proposal for ECMAScript, and is an influence and inspiration for my work as a co-champion of the [Structs and Shared Structs Proposal](https://github.com/tc39/proposal-structs).

Finally, there are a number of utility packages that are also based on future proposals, or existing functionality in ECMAScript:
- `@esfx/ref` is a rough approximation of some of what I've been considering in my [`ref` Proposal](https://github.com/rbuckton/proposal-refs)
- `@esfx/indexed-object` is a rough approximation of the existing "Integer Indexed Exotic Object" in the ECMAScript specification that is the basic for how numeric indexing works for `Array`s and Typed Arrays (i.e., `Uint32Array`, etc.).

There are more than a few other packages I haven't covered in this list. Many are shims to add the low-level protocols I mentioned above onto existing built-ins, or adapters to provide interoperability for packages like `@esfx/async-canceltoken` with the DOM's `AbortController`/`AbortSignal`.

**Reply by hinell** [reply]
Nice. Now put this into Readme as everyone can learn. Thanks.

What does the _protocol_ mean btw?

**Reply by rbuckton** [reply]
Everything I mentioned above is covered at https://esfx.js.org/, which is the homepage link in the explainer.

"Protocol" in this sense means something similar to "interface". The "protocols" proposal chose its name to avoid a conflict with the term "interface" as used by TypeScript. Essentially it establishes a set of pre-defined `symbol`s whose intent is to convey implementation of a specific behavior, much as a `Symbol.iterator`  method indicates you intend to implement the `Iterable` interface, and its result is an `Iterator`, which establishes an iteration protocol.

**Comment by unicomp21** [comment]
Lol, it's like the rest of the planet is walking everywhere when we have a
Mazerati right here!  lol
